File: query_primer/query_primer__square_division.cs
using System.Linq;
using static System.Console;
using static System.Math;

internal class Program
{
    static void Main()
    {
        const int num = 10000;
        int k = int.Parse( ReadLine() );
        int sq = (int)Sqrt( num );
        int[] a = new int[num];
        int[] am = new int[sq];

        SetOut( new System.IO.StreamWriter( OpenStandardOutput() ) { AutoFlush = false } );
        foreach( int i in Enumerable.Range( 0, num ) ) {
            a[i] = int.Parse( ReadLine() );
        }
        foreach( int i in Enumerable.Range( 0, sq ) ) {
            foreach( int j in Enumerable.Range( 0, sq ) ) {
                am[i] = j == 0 ? a[sq * i] : Max( am[i], a[sq * i + j] );
            }
        }
        foreach( int _ in Enumerable.Range( 0, k ) ) {
            int i = 0;
            int o = 0;
            int[] lr = ReadLine().Split().Select( x => int.Parse( x ) - 1 ).ToArray();

            i = lr[0];
            o = a[lr[0]];
            while( i <= lr[1] ) {
                if( i % sq == 0 && i + sq - 1 <= lr[1] ) {
                    o = Max( o, am[i / sq] );
                    i += sq;
                } else {
                    o = Max( o, a[i] );
                    i++;
                }
            }
            WriteLine( o );
        }
        Out.Flush();
        return;
    }
}

File: query_primer/query_primer__square_division.py
import math

k = int(input())
N = 10000
s = math.floor(math.sqrt(N))
a = [0 for _ in range(N)]
m = [0 for _ in range(s)]

for i in range(N):
    a[i] = int(input())
for i in range(s):
    for j in range(s):
        m[i] = a[s * i] if j == 0 else max(m[i], a[s * i + j])
for _ in range(k):
    l, r = list(map(int, input().split()))
    l -= 1
    r -= 1
    i = l
    o = a[l]

    while i <= r:
        if i % s == 0 and i + s - 1 <= r:
            o = max(o, m[i // s])
            i += s
        else:
            o = max(o, a[i])
            i += 1
    print(o)
  # 実行時間 2570ms
